compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +442,234 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
+          if (Operation *operandOp = v.getDefiningOp())
+            assert(!regionOp->isAncestor(operandOp) &&
+                   "could not infer result shape");
+        }
+      }
+#endif  // NDEBUG
+      resultDims.push_back(dimValue);
+      assert(resultDims.back().isa<BlockArgument>() ||
+             !regionOp->isAncestor(resultDims.back().getDefiningOp()) &&
+                 "result dim defined in containing op");
+    }
+  }
+
+  // Find tied results.
+  DenseSet<Value> tiedArgumentsSet;
+  SmallVector<int64_t> tiedArguments(numResults,
+                                     IREE::Util::TiedOpInterface::kUntiedIndex);
+  for (const auto &it : llvm::enumerate(terminator->getOperands())) {
+    auto tiedArgument =
+        findFirstTiedValueOutsideOfContainingOp(regionOp, it.value());
+    if (!tiedArgument.hasValue()) continue;
+    assert(argumentsSet.contains(*tiedArgument) &&
+           "expected that tiedArgument is already an argument");
+    // Do not tie an argument to multiple results.
+    if (tiedArgumentsSet.contains(*tiedArgument)) continue;
+    tiedArgumentsSet.insert(*tiedArgument);
+    tiedArguments[it.index()] = std::distance(
+        argumentsSet.begin(), llvm::find(argumentsSet, *tiedArgument));
+  }
+
+  // Create empty dispatch region.
+  SmallVector<Value> arguments(argumentsSet.begin(), argumentsSet.end());
+  arguments.append(argumentDims);
+  arguments.append(resultDims);
+  auto workgroupsOp = rewriter.create<IREE::Flow::DispatchWorkgroupsOp>(
+      loc, /*workload=*/ValueRange(), resultTypes, resultDims, arguments,
+      argumentDims, tiedArguments);
+  BlockAndValueMapping bvm;
+  bvm.map(arguments, workgroupsOp.getInputBlockArguments());
+
+  // Create DispatchTensorLoadOp for all tensor arguments.
+  assert(workgroupsOp.getWorkgroupBody().hasOneBlock() &&
+         "expected one block after constructor");
+  Block &newBody = workgroupsOp.getWorkgroupBody().getBlocks().front();
+  assert(newBody.empty() && "expected empty block after constructor");
+  rewriter.setInsertionPointToStart(&newBody);
+  for (const auto &it : llvm::enumerate(arguments)) {
+    auto tensorType = it.value().getType().dyn_cast<RankedTensorType>();
+    if (!tensorType) continue;
+    auto inputBbArg = workgroupsOp.getInputBlockArgument(it.index());
+    auto dims =
+        Util::findVariadicDynamicDims(it.index(), arguments, argumentDims);
+    assert(dims.size() == tensorType.getNumDynamicDims() &&
+           "dynamic dims not found among arguments");
+    SmallVector<Value> bbArgDims = llvm::to_vector(
+        llvm::map_range(dims, [&](Value v) { return bvm.lookup(v); }));
+    Value loadedTensor = rewriter.create<IREE::Flow::DispatchTensorLoadOp>(
+        loc, tensorType, inputBbArg, bbArgDims);
+    bvm.map(it.value(), loadedTensor);
+  }
+
+  // Clone contents of the containing op into the dispatch region.
+  // TODO: Move instead of clone for better performance.
+  for (Operation &op : body) rewriter.clone(op, bvm);
```

Comment:
Could we just splice the statements from the `flow.dispatch.region` operation into the `flow.dispatch.workgroup` operation? You can just use `replaceUsesWithIf` to replace all uses of the captured value within the block to bbArg. Cloning can be expensive, this effectively clones the whole program....

Reply:
Ah yes this works.